src/replication/replication_object.cpp

```diff
@@ -835,4 +836,77 @@ namespace cubreplication
     m_data_size += rec_size;
   }
 
+  savepoint_object::savepoint_object (const char *savepoint_name, event_type event)
+    : m_savepoint_name (savepoint_name)
+    , m_event (event)
+  {
+  }
+
+  int
+  savepoint_object::apply ()
+  {
+    cubthread::entry &thread_ref = cubthread::get_entry ();
+    if (m_event == CREATE_SAVEPOINT)
+      {
+	if (log_append_savepoint (&thread_ref, m_savepoint_name.c_str ()) == NULL)
+	  {
+	    assert (false);
+	    return ER_FAILED;
+	  }
+      }
+    else
+      {
+	LOG_LSA savept_lsa;
+	if (log_abort_partial (&thread_ref, m_savepoint_name.c_str (), &savept_lsa) != TRAN_UNACTIVE_ABORTED)
+	  {
+	    assert (false);
+	    return ER_FAILED;
+	  }
+      }
+    return NO_ERROR;
+  }
+
+  void
+  savepoint_object::stringify (string_buffer &str)
+  {
+    if (m_event == CREATE_SAVEPOINT)
+      {
+	str ("create ");
+      }
+    else
+      {
+	str ("rollback to ");
+      }
+    str ("savepoint ");
+    str (m_savepoint_name.c_str ());
+  }
+
+  void
+  savepoint_object::pack (cubpacking::packer &serializator) const
+  {
+    serializator.pack_int (PACKING_ID);
+    serializator.pack_string (m_savepoint_name);
+    serializator.pack_to_int (m_event);
+  }
+
+  void
+  savepoint_object::unpack (cubpacking::unpacker &deserializator)
+  {
+    int check_id;
+    deserializator.unpack_int (check_id);
+    assert (check_id == PACKING_ID);
+    deserializator.unpack_string (m_savepoint_name);
+    deserializator.unpack_from_int (m_event);
+  }
+
+  std::size_t
+  savepoint_object::get_packed_size (cubpacking::packer &serializator, std::size_t start_offset) const
+  {
+    size_t size = 0;
+    size += serializator.get_packed_int_size (start_offset + size);
+    size += serializator.get_packed_string_size (m_savepoint_name, start_offset + size);
```

Comment:
start_offset is already included in size after first "pack" call.

Reply:
Please explain what you mean. As far as I understand, get packed size function will only return the size of one item (optionally given as argument) with alignment padding.

Reply:
Existing "get_packed_size" functions seems to use total size (they start with offset).